### ISSUE TITLE
(aws-iam): policy document optimization

### ISSUE THREAD

**Issue by andreialecu** [aws/aws-cdk#14713]
I wonder if this is an acceptable feature request. If so, I could try submitting a PR myself.

`aws-iam` currently creates sub-optimal policies that could be pretty easily optimized. For example consider a helper method that runs this on an ECR repository:

```ts
    policy.addToPolicy(
      new iam.PolicyStatement({
        resources: [repository.repositoryArn],
        actions: [
          "ecr:InitiateLayerUpload",
          "ecr:UploadLayerPart",
          "ecr:CompleteLayerUpload",
          "ecr:BatchCheckLayerAvailability",
          "ecr:PutImage",
        ],
      })
    );
```

If `policy` is a "shared" policy, and the stack adds new ECR repositories over time, this construct will very quickly run into #11562.

The optimization would be that instead of creating duplicate statements with the same actions for every separate resource, it could just create one policy statement and merge all the resources together.

For example, it would be the equivalent of adding just one policy of:

```ts
    policy.addToPolicy(
      new iam.PolicyStatement({
        resources: [repository.repositoryArn, repository2.repositoryArn, ...etc],
        actions: [
          "ecr:InitiateLayerUpload",
          "ecr:UploadLayerPart",
          "ecr:CompleteLayerUpload",
          "ecr:BatchCheckLayerAvailability",
          "ecr:PutImage",
        ],
      })
    );
```

Note that I'm aware that new resources could be added to the `PolicyStatement` instead of adding new statements to the policy, but this results in a lot of extra complexity in the stack's implementation, and I feel that cdk could do an optimization here instead.

Additionally, there are places like: https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-iam/lib/role.ts#L241-L246 which are outside of the user's control, which suffer from the same problem (this very quickly runs into https://github.com/aws/aws-cdk/issues/11562#issuecomment-824025826) but which could be optimized by this step.

### Use Case

The main issue is that the resulting CloudFormation template ends up huge, and runs over some internal limits as mentioned in https://github.com/aws/aws-cdk/issues/11562#issuecomment-787134785 and https://github.com/aws/aws-cdk/issues/11562#issuecomment-824025826

### Proposed Solution

My initial idea is to add a post-processor similar to the one in:
https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-iam/lib/policy-document.ts#L55

It would detect if the generated policy can be optimized by merging similar resources or actions together.

### Other

I think this shouldn't be a breaking change if implemented correctly. Also I'm not sure if it requires any changes, or if it affects the `IAM Statement Changes` interactive prompt.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by andreialecu** [comment]
I implemented this in https://github.com/aws/aws-cdk/pull/14714

**Comment by rix0rrr** [comment]
Leaving this open for future upvotes which may increase the weight, but we consciously decided not to do anything about this.

IAM policy optimization is attractive but also dangerous, and we're super scared of messing up.

If you want to implement this for now, I would suggest you implement an Aspect to twiddle the construct tree out-of-band.

**Comment by andreialecu** [comment]
@rix0rrr wondering if you saw my comment just above you? The PR received no comments.

**Comment by bilalq** [comment]
The absence of this results in AWS CodePipelines being unusable for cross-region deployments in many cases. See #16244 for more details.

**Comment by comcalvi** [comment]
I think this has been resolved by #19114
